get_fstar_z3.sh

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+kernel="$(uname -s)"
+case "$kernel" in
+  CYGWIN*) kernel=Windows ;;
+esac
+
+arch="$(uname -m)"
+case "$arch" in
+  arm64) arch=aarch64 ;;
+esac
+
+declare -A release_url
+release_url["Linux-x86_64-4.8.5"]="https://github.com/Z3Prover/z3/releases/download/Z3-4.8.5/z3-4.8.5-x64-ubuntu-16.04.zip"
+release_url["Darwin-x86_64-4.8.5"]="https://github.com/Z3Prover/z3/releases/download/Z3-4.8.5/z3-4.8.5-x64-osx-10.14.2.zip"
+release_url["Windows-x86_64-4.8.5"]="https://github.com/Z3Prover/z3/releases/download/Z3-4.8.5/z3-4.8.5-x64-win.zip"
+release_url["Linux-x86_64-4.13.3"]="https://github.com/Z3Prover/z3/releases/download/z3-4.13.3/z3-4.13.3-x64-glibc-2.35.zip"
+release_url["Linux-aarch64-4.13.3"]="https://github.com/Z3Prover/z3/releases/download/z3-4.13.3/z3-4.13.3-arm64-glibc-2.34.zip"
+release_url["Darwin-x86_64-4.13.3"]="https://github.com/Z3Prover/z3/releases/download/z3-4.13.3/z3-4.13.3-x64-osx-13.7.zip"
+release_url["Darwin-aarch64-4.13.3"]="https://github.com/Z3Prover/z3/releases/download/z3-4.13.3/z3-4.13.3-arm64-osx-13.7.zip"
+release_url["Windows-x86_64-4.13.3"]="https://github.com/Z3Prover/z3/releases/download/Z3-4.8.5/z3-4.8.5-x64-win.zip"
+
+trap "exit 1" HUP INT PIPE QUIT TERM
+cleanup() {
+  if [ -n "${tmp_dir:-}" ]; then
+    rm -rf "$tmp_dir"
+  fi
+}
+trap "cleanup" EXIT
+
+download_z3() {
+  url="$1"
+  version="$2"
+  destination_file_name="$3"
+
+  if [ -z "${tmp_dir:-}" ]; then
+    tmp_dir="$(mktemp -d --tmpdir get_fstar_z3.XXXXXXX)"
+  fi
+
+  echo ">>> Downloading Z3 $version from $url ..."
+  base_name="$(basename "$url")"
+
+  z3_path="${base_name%.zip}/bin/z3"
+  if [ "$kernel" = Windows ]; then z3_path="$z3_path.exe"; fi
+
+  pushd "$tmp_dir"
+  curl -L "$url" -o "$base_name"
+  unzip "$base_name" "$z3_path"
+  popd
+
+  install -m0755 "$tmp_dir/$z3_path" "$destination_file_name"
+  echo ">>> Installed Z3 $version to $destination_file_name"
+}
+
+dest_dir="$1"
+if [ -z "$dest_dir" ]; then
+  echo "Usage: get_fstar_z3.sh destination/directory/bin"
+  exit 1
+fi
+
+mkdir -p "$dest_dir"
+
+for z3_ver in 4.8.5 4.13.3; do
+  key="$kernel-$arch-$z3_ver"
+  url="${release_url[$key]:-}"
+
+  destination_file_name="$dest_dir/z3-$z3_ver"
+  if [ "$kernel" = Windows ]; then destination_file_name="$destination_file_name.exe"; fi
+
+  if [ -f "$destination_file_name" ]; then
+    echo ">>> Z3 $z3_ver already downloaded to $destination_file_name"
+  elif [ -z "$url" ]; then
+    echo ">>> Z3 $z3_ver not available for this architecture, skipping..."
```

Comment:
On Apple Silicon, this declines to install 4.8.5. However, 4.8.5 can run on OSX, via its x64 emulation facilities (a.k.a. Rosetta 2). This is actually required for HACL* and other projects which have not yet upgraded to the latest z3.

I would suggest special-casing 4.8.5 / aarch64 to download the x64 version, and printing out a notice that Rosetta 2 ought to be installed on this machine.

A few more remarks:
- you'll probably have a better time using [[ for you tests, since this is lexed and parsed by bash itself, meaning it can deal with variables that expand to empty, rather than [, which has to obey command-line argument parsing as if it were calling the actual /usr/bin/[, and therefore will yield unexpected errors if owing to any circumstances some of your variables are actually empty
- you could use the `local` keyword in a few of your functions for local variables
- your script requires bash >= 4 (associative arrays) but does not check for it -- this is pertinent for OSX, which ships bash 3 (the last GPLv2 version), but not bash 4; note that for now this is probably fine since this is only ever called via the everest script, which does run proper checks.

Reply:
Thanks! See https://github.com/FStarLang/FStar/pull/3645

Reply:
> you'll probably have a better time using [[ for you tests, since this is lexed and parsed by bash itself, meaning it can deal with variables that expand to empty, rather than [, which has to obey command-line argument parsing as if it were calling the actual /usr/bin/[, and therefore will yield unexpected errors if owing to any circumstances some of your variables are actually empty

`[` works just fine with empty variables.  You need to quote variables, but you should always quote them anyhow.

> your script requires bash >= 4 (associative arrays) but does not check for it -- this is pertinent for OSX, which ships bash 3

I didn't test it at all on macos, thanks for serving as the guinea pig!  fstarlang/fstar#3646 (I see Guido beat me to it.)